### PR TITLE
[Frontend] Tool calling parser for Granite 3.0 models

### PR DESCRIPTION
**Pull request by maxdebayser**
This PR adds a tool calling parser for `ibm-granite/granite-3.0-8b-instruct`. The smaller models in 
the [Granite 3.0 Language Models](https://huggingface.co/collections/ibm-granite/granite-30-models-66fdb59bbb54785c3512114f) models collection pass some of the tests..

cc: @njhill

### CODE REVIEW

File: docs/source/serving/openai_compatible_server.md
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:160]: @maxdebayser should `granite` also be added here?

File: docs/source/serving/openai_compatible_server.md
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:233]:
```suggestion
* `ibm-granite/granite-3.0-8b-instruct`
```

File: docs/source/serving/openai_compatible_server.md
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:231]: Is `--chat-template` missing here? (see llama above for example)

File: docs/source/serving/openai_compatible_server.md
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:236]: Is `--chat-template` missing here?

File: docs/source/serving/openai_compatible_server.md
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:258]: Duplicate, should be removed?

File: docs/source/serving/openai_compatible_server.md
Comment (njhill) [inline docs/source/serving/openai_compatible_server.md:228]: Probably better to swap the order of these and move the granite 3.0 one to the top

**Review by njhill — COMMENTED** [review]
Thanks @maxdebayser!

**Comment by maxdebayser** [comment]
This is ready for review.

cc: @njhill @wseaton

File: vllm/entrypoints/openai/tool_parsers/granite_tool_parser.py
Comment (njhill) [inline vllm/entrypoints/openai/tool_parsers/granite_tool_parser.py:204]: nit: remove this empty line?

**Review by njhill — COMMENTED** [review]
Thanks @maxdebayser! I did a first pass and left some comments

**Comment by njhill** [comment]
cc @K-Mistele in case you'd like to take a look too

**Comment by K-Mistele** [comment]
will take another look!

**Review by njhill — APPROVED** [review]
Thanks again @maxdebayser. I'll hold off for any comments from @K-Mistele before merging.

**Comment by K-Mistele** [comment]
Giving it a final once-over :) 

Out of curiosity it looks like the model's default context is 4096; is there any way to scale this if I want to test longer-context tool calls? no worries if there's not a good way to do with with vLLM

**Comment by K-Mistele** [comment]
Looks great to me! Seems pretty robust, and passing tests on my machine :)

**Comment by K-Mistele** [comment]
Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337

**Comment by DarkLight1337** [comment]
> Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337

The test is not failing on main branch, so I think it's introduced by this PR. Do tell me if I'm mistaken though.

**Comment by K-Mistele** [comment]
> > Looks like the only failing tests were AMD-related which was expected last I checked. cc @mgoin @DarkLight1337
> 
> The test is not failing on main branch, so I think it's introduced by this PR. Do tell me if I'm mistaken though.

Oops, missed that. I just remembered the last time I did one there were some expected to fail, but I'm sure you're right :) 

Seems related to the FP8 granite 20B quantization (`mbayser/granite-20b-functioncalling-FP8-KV`) that's being used:

https://buildkite.com/vllm/ci-aws/builds/10840#01930511-9576-45cb-aebb-e237d5f07c9b/974-5501

AMD supports FP8 in ROCM > 6.2. Could also be an OOM issue since this is a 20B that we had issues fitting into other CI GPUs; or could be related to CPU offloading. 

Whatever the case, seems more like a model+configuration+hardware compatibility issue rather than a code issue. Is there any way to disable this particular model for AMD tests, and would that be an acceptable solution?

**Comment by DarkLight1337** [comment]
I think this is alright. @njhill any thoughts?

**Comment by maxdebayser** [comment]
I've pushed a change to disable skip the `granite-20b-functioncalling` test on AMD. Let's see if it passes now :crossed_fingers:

**Comment by maxdebayser** [comment]
> Out of curiosity it looks like the model's default context is 4096; is there any way to scale this if I want to test longer-context tool calls? no worries if there's not a good way to do with with vLLM

I'm not sure, but as we add more models and tests, we might have to get the max sequence length for each model and skip the tests that require longer ones.

**Comment by njhill** [comment]
Thanks again @maxdebayser @K-Mistele @DarkLight1337!

**Pull request merged**
